### PR TITLE
Change: Reunify Version based policy split

### PR DESCRIPTION
**Pull request by nickanderson**
Version based policy splits were introduced when the masterfiles policy
framework started using new functionaly that was incompatible with old
versions. For example a function introduced in version 3.7 would be
unknown and generate a syntax error for 3.6. 3.7 introduced the @if
macro which allows policy to be guarded based on a minimum version of
cfengine. This functionality allows us to re-unify the version based
split for 3.7+ clients.

Changelog: Title

- These files were removed in favor of a more recent version
  - deleted:    controls/3.7/cf_agent.cf
  - deleted:    controls/3.7/cf_hub.cf
  - deleted:    controls/3.7/cf_monitord.cf
  - deleted:    controls/3.7/cf_runagent.cf
  - deleted:    controls/3.7/cf_serverd.cf
  - deleted:    controls/3.7/def.cf
  - deleted:    controls/3.7/def_inputs.cf
  - deleted:    controls/3.7/reports.cf
  - deleted:    controls/3.7/update_def.cf
  - deleted:    controls/3.7/update_def_inputs.cf
  - deleted:    controls/3.8/cf_execd.cf
  - deleted:    lib/3.7/autorun.cf
  - deleted:    lib/3.7/bundles.cf
  - deleted:    lib/3.7/cfe_internal.cf
  - deleted:    lib/3.7/cfengine_enterprise_hub_ha.cf
  - deleted:    lib/3.7/commands.cf
  - deleted:    lib/3.7/common.cf
  - deleted:    lib/3.7/databases.cf
  - deleted:    lib/3.7/edit_xml.cf
  - deleted:    lib/3.7/examples.cf
  - deleted:    lib/3.7/feature.cf
  - deleted:    lib/3.7/files.cf
  - deleted:    lib/3.7/guest_environments.cf
  - deleted:    lib/3.7/monitor.cf
  - deleted:    lib/3.7/packages.cf
  - deleted:    lib/3.7/paths.cf
  - deleted:    lib/3.7/processes.cf
  - deleted:    lib/3.7/services.cf
  - deleted:    lib/3.7/stdlib.cf
  - deleted:    lib/3.7/storage.cf
  - deleted:    lib/3.7/users.cf
  - deleted:    lib/3.7/vcs.cf
  - deleted:    lib/3.8/cfe_internal_hub.cf

- These files were moved to the unified lib directory
  - renamed:    controls/3.8/cf_agent.cf -> controls/cf_agent.cf
  - renamed:    controls/3.7/cf_execd.cf -> controls/cf_execd.cf
  - renamed:    controls/3.8/cf_hub.cf -> controls/cf_hub.cf
  - renamed:    controls/3.8/cf_monitord.cf -> controls/cf_monitord.cf
  - renamed:    controls/3.8/cf_runagent.cf -> controls/cf_runagent.cf
  - renamed:    controls/3.8/cf_serverd.cf -> controls/cf_serverd.cf
  - renamed:    controls/3.8/def.cf -> controls/def.cf
  - renamed:    controls/3.8/def_inputs.cf -> controls/def_inputs.cf
  - renamed:    controls/3.8/reports.cf -> controls/reports.cf
  - renamed:    controls/3.8/update_def.cf -> controls/update_def.cf
  - renamed:    controls/3.8/update_def_inputs.cf -> controls/update_def_inputs.cf
  - renamed:    lib/3.8/autorun.cf -> lib/autorun.cf
  - renamed:    lib/3.8/bundles.cf -> lib/bundles.cf
  - renamed:    lib/3.8/cfe_internal.cf -> lib/cfe_internal.cf
  - renamed:    lib/3.7/cfe_internal_hub.cf -> lib/cfe_internal_hub.cf
  - renamed:    lib/3.8/cfengine_enterprise_hub_ha.cf -> lib/cfengine_enterprise_hub_ha.cf
  - renamed:    lib/3.8/commands.cf -> lib/commands.cf
  - renamed:    lib/3.8/common.cf -> lib/common.cf
  - renamed:    lib/3.8/databases.cf -> lib/databases.cf
  - renamed:    lib/3.8/edit_xml.cf -> lib/edit_xml.cf
  - renamed:    lib/3.8/examples.cf -> lib/examples.cf
  - renamed:    lib/3.8/feature.cf -> lib/feature.cf
  - renamed:    lib/3.8/files.cf -> lib/files.cf
  - renamed:    lib/3.8/guest_environments.cf -> lib/guest_environments.cf
  - renamed:    lib/3.8/monitor.cf -> lib/monitor.cf
  - renamed:    lib/3.8/packages.cf -> lib/packages.cf
  - renamed:    lib/3.8/paths.cf -> lib/paths.cf
  - renamed:    lib/3.8/processes.cf -> lib/processes.cf
  - renamed:    lib/3.8/services.cf -> lib/services.cf
  - renamed:    lib/3.8/stdlib.cf -> lib/stdlib.cf
  - renamed:    lib/3.8/storage.cf -> lib/storage.cf
  - renamed:    lib/3.8/users.cf -> lib/users.cf
  - renamed:    lib/3.8/vcs.cf -> lib/vcs.cf

- These files were modified to support the version re-unification.
  - modified:   promises.cf

### CODE REVIEW

File: promises.cf [hunk 1]
```diff
@@ -208,9 +207,20 @@ bundle common cfengine_stdlib
...
+      # errors introduced by new functionality. For example new functions.
+      # This also works for 3.8 because local_libdir should be set to lib
+      # instead of lib/3.8
       "inputs" slist => { "$(sys.local_libdir)/stdlib.cf" };
 
+    cfengine_3_7::
+      # CFEngine 3.7 has local_libdir set to $(sys.inputdir)/lib/3.7, but with
+      # the @if macro support we can re-unify the split library for 3.7+ so we
+      # specify the unified lib instead of local_libdir.
+      "inputs" slist => { "lib/stdlib.cf" };
```
Comment (tzz) [inline promises.cf]: This won't always work.  Instead, I would use `$(sys.local_libdir)/..` with the understanding that it's a temporary workaround.

Comment (tzz) [inline promises.cf]: I can't recall the error cases, unfortunately, but I remember running into issues with relative paths.  Sorry if that's not helpful.

**Comment by tzz** [comment]
It's a good change.  I think `lib/3.5` should be removed now and `lib/README.md` should be updated so it doesn't mention the old `libraries/cfengine_stdlib.cf`.

**Comment by nickanderson** [comment]
Regarding lib/3.5 we try to keep 3 versions back masterfiles compatibility so we can't drop it yet. This is really only there to support upgrades from 3.5.

Regarding libraries/cfengine_stdlib.cf I agree

**Comment by nickanderson** [comment]
HUM the library split does not work with 3.5.0? ... according to the readme

**Comment by nickanderson** [comment]
However the only reference I can find to cfengine_stdlib.cf is in the readme and a comment in one of the 3.5 stdlib files. This seems incorrect, will test.

**Comment by nickanderson** [comment]
After talking with Eystein, 3.5 support should be removed

**Comment by nickanderson** [comment]
trigger build

**Comment by nickanderson** [comment]
ping @estenberg

**Comment by estenberg** [comment]
Looks good to me! 16k lines removed -- what's not to like? :+1:

**Comment by nickanderson** [comment]
Merging then

**Comment by tzz** [comment]
I think this will be a good change in the long term.  Thanks a lot for making it!

**Comment by kacfengine** [comment]
@nickanderson: Great! So I guess you can close this ticket now? https://dev.cfengine.com/issues/7252

Note that the build fails with this change, since the tests are not adapted. Can you fix it according to the new structure? This is one example, http://jenkins.usdc.cfengine.com/job/testing-enterprise-master/lastCompletedBuild/label=PACKAGES_x86_64_linux_SuSE_11/testReport/._promises_promises/cf/promises_cf/, and there are many more like it.

**Comment by kacfengine** [comment]
Don't forget the unsafe ones!

**Comment by nickanderson** [comment]
@kacfengine I think I will need help to figure out where those test failures are coming from. There is a makefile in acceptance that has a reference to masterfiles/lib/3.7. Do you think thats what might be causing this? https://github.com/cfengine/core/blob/master/tests/acceptance/Makefile.am#L95

**Comment by kacfengine** [comment]
No, that part is only run on user demand. It will also need fixing most likely, but that is less urgent.

I bet the failure is the test referring to the (re-)moved policy files through ``inputs``. Check the ``inputs`` attribute of the test I mentioned, for example. Most likely they will need to be updated to the new location.

``make check`` should show you the error btw, since it's not only unsafe tests failing.

**Comment by nickanderson** [comment]
@kacfengine i get errors trying to build master locally.

```
> $ make                                                                        [±master ●▴]
Making all in libcompat
  CC       generic_at.lo
In file included from ../libutils/platform.h:358:0,
                 from generic_at.c:25:
/usr/include/linux/if.h:71:2: error: redeclaration of enumerator 'IFF_UP'
  IFF_UP    = 1<<0,  /* sysfs */
```

When I `git grep stdlib.cf` in the root of the repo I don't see anything that says 3.8 I do see some references in examples to `$(sys.libdir)/stdlib.cf` but `sys.libdir` out to be set to `lib` not `lib/3.8` in master.

**Comment by tzz** [comment]
Oh, I know this one @nickanderson I just went through the same pain.   autoconf sucks sometimes.

Use `git clean -f` to really clean your whole checkout, then run `autogen.sh` etc.

**Comment by nickanderson** [comment]
THANKS! helped me figure out I was missing libtool

**Comment by nickanderson** [comment]
So I can build now.  Still struggling to see exactly where the failure is from.

I do get a bunch of failed tests, but if I run them directly they pass:

Using testall --bindir=/tmp/cfengine/bin (where I installed my build)
```
./01_vars/03_lists/014.cf FAIL (UNEXPECTED FAILURE)
```

vs running directly
```
> $ /var/cfengine/bin/cf-agent -KIf ./01_vars/03_lists/014.cf -D AUTO,DEBUG      [±master ▴]
    info: Created file '/tmp/TEST.cfengine.actual', mode 0600
    info: Edit file '/tmp/TEST.cfengine.actual'
    info: Created file '/tmp/TEST.cfengine.expected', mode 0600
    info: Edit file '/tmp/TEST.cfengine.expected'
    info: Executing 'no timeout' ... '/usr/bin/env LC_ALL=C /usr/bin/sort /tmp/TEST.cfengine.actual > /tmp/TEST.cfengine.actual.sorted'
    info: Completed execution of '/usr/bin/env LC_ALL=C /usr/bin/sort /tmp/TEST.cfengine.actual > /tmp/TEST.cfengine.actual.sorted'
    info: Executing 'no timeout' ... '/usr/bin/env LC_ALL=C /usr/bin/sort /tmp/TEST.cfengine.expected > /tmp/TEST.cfengine.expected.sorted'
    info: Completed execution of '/usr/bin/env LC_ALL=C /usr/bin/sort /tmp/TEST.cfengine.expected > /tmp/TEST.cfengine.expected.sorted'
R: /home/nickanderson/CFEngine/core/tests/acceptance/./01_vars/03_lists/014.cf Pass
```
The failure you mention says:
```
error: Can"t stat file "../../lib/3.8/stdlib.cf" for parsing. (stat: No such file or directory)
```

Hum, wait it also says:

```
masterfiles/tests/acceptance/./promises/promises.cf FAIL
```
So im guessing this test is in masterfiles, not core where I was looking.

**Comment by nickanderson** [comment]
Testing from masterfiles repo:
```
> $ ./testall --agent=/tmp/cfengine/bin/cf-agent                                        [±master ●▴]
...
./promises/autorun.cf FAIL
./promises/autorun-D.cf FAIL
./promises/promises.cf FAIL
```
But when I run one directly:

```
> $ /tmp/cfengine/bin/cf-agent -V                                                       [±master ●▴]
CFEngine Core 3.8.0a1.4486dd3

> $ /tmp/cfengine/bin/cf-agent -KIf ./promises/promises.cf -D AUTO,DEBUG                [±master ●▴]
    info: Executing 'no timeout' ... '"/home/nickanderson/.cfagent/bin/cf-promises" -c --eval-functions -f ../../promises.cf'
    info: Completed execution of '"/home/nickanderson/.cfagent/bin/cf-promises" -c --eval-functions -f ../../promises.cf'
R: /home/nickanderson/CFEngine/masterfiles/tests/acceptance/./promises/promises.cf Pass
```

When I "git grep stdlib.cf" all references to stdlib.cf use sys.local_libdir

However, it seems that libdir is using a version specific path still, and it out not. I think this is a core change that still needs to take place.
```
 > $ /tmp/cfengine/bin/cf-promises -cf ./acceptance/promises/promises.cf --show-vars | grep libdir
default:sys.libdir                       /home/nickanderson/.cfagent/inputs/lib/3.8                   source=agent                            
default:sys.local_libdir                 lib/3.8                                                      source=agent                            
```

**Comment by nickanderson** [comment]
I_dont_know_C::
  https://github.com/cfengine/core/blob/master/libenv/sysinfo.c#L394-L398

**Comment by nickanderson** [comment]
I think this patch against core would fix it: 

```
diff --git a/libenv/sysinfo.c b/libenv/sysinfo.c
index a75c6aa..a467221 100644
--- a/libenv/sysinfo.c
+++ b/libenv/sysinfo.c
@@ -391,13 +391,13 @@ void DiscoverVersion(EvalContext *ctx)
         snprintf(workbuf, CF_MAXVARSIZE, "%d", patch);
         EvalContextVariablePutSpecial(ctx, SPECIAL_SCOPE_SYS, "cf_version_patch", workbuf, CF_DATA_TYPE_STRING, "source=agent");
 
-        snprintf(workbuf, CF_BUFSIZE, "%s%cinputs%clib%c%d.%d",
+        snprintf(workbuf, CF_BUFSIZE, "%s%cinputs%clib",
                  workdir, FILE_SEPARATOR, FILE_SEPARATOR,
                  FILE_SEPARATOR, major, minor);
 
         EvalContextVariablePutSpecial(ctx, SPECIAL_SCOPE_SYS, "libdir", workbuf, CF_DATA_TYPE_STRING, "source=agent");
 
-        snprintf(workbuf, CF_BUFSIZE, "lib%c%d.%d", FILE_SEPARATOR, major, minor);
+        snprintf(workbuf, CF_BUFSIZE, "lib", FILE_SEPARATOR, major, minor);
         EvalContextVariablePutSpecial(ctx, SPECIAL_SCOPE_SYS, "local_libdir", workbuf, CF_DATA_TYPE_STRING, "source=agent");
     }
     else
```

```
> $ /tmp/cfengine/bin/cf-promises -cf ./promises.cf --show-vars | grep libdir           [±master ●▴]
default:sys.libdir                       /home/nickanderson/.cfagent/inputs/lib                       source=agent                            
default:sys.local_libdir                 lib                                                          source=agent       
```

PR: https://github.com/cfengine/core/pull/2318

**Pull request merged**